tsdb/importer/import_test.go

```diff
@@ -0,0 +1,753 @@
+		str := fmt.Sprintf("# HELP %s This is a metric\n# TYPE %s %s", metricName, metricName, metricType)
+		for _, s := range series {
+			str += fmt.Sprintf("\n%s%s %f %d", metricName, s.Labels.String(), s.Value, s.Timestamp)
+		}
+		return str
+	}
+	metricA := "metricA"
+	metricLabels := []string{"foo", "bar"}
+	mint, maxt := int64(0), int64(1000)
+	step := 5
+	seriesA := genSeries(metricLabels, mint, maxt, step)
+	textA := partialOMText(metricA, "gauge", seriesA)
+	metricB := "metricB"
+	seriesB := genSeries(metricLabels, mint, maxt, step)
+	textB := partialOMText(metricB, "gauge", seriesB)
+
+	text := fmt.Sprintf("%s\n%s\n# EOF", textA, textB)
+
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	testutil.Ok(t, err)
+	err = ImportFromFile(strings.NewReader(text), tmpDbDir, maxSamplesInMemory, maxBlockChildren, nil)
+	testutil.Ok(t, err)
+
+	addMetricLabel := func(series []tsdb.MetricSample, metricName string) []tsdb.MetricSample {
+		augSamples := make([]tsdb.MetricSample, 0)
+		for _, sample := range series {
+			lbls := labels2.FromStrings("__name__", metricName)
+			s := tsdb.MetricSample{
+				Timestamp: sample.Timestamp,
+				Value:     sample.Value,
+				Labels:    append(lbls, sample.Labels...),
+			}
+			augSamples = append(augSamples, s)
+		}
+		return augSamples
+	}
+
+	expectedSamples := append(addMetricLabel(seriesA, metricA), addMetricLabel(seriesB, metricB)...)
+	expectedSymbols := append([]string{"__name__", metricA, metricB}, metricLabels...)
+
+	db, err := tsdb.Open(tmpDbDir, nil, nil, &tsdb.Options{
+		RetentionDuration:      tsdb.DefaultOptions().RetentionDuration,
+		AllowOverlappingBlocks: true,
+	})
+	testutil.Ok(t, err)
+
+	blocks := make([]tsdb.BlockReader, 0)
+	for _, b := range db.Blocks() {
+		blocks = append(blocks, b)
+	}
+
+	testBlocks(t, blocks, metricLabels, mint, maxt-int64(step-1), expectedSamples, expectedSymbols, 1)
+
+	_ = os.RemoveAll(tmpDbDir)
+}
+
+// Test to check if we can detect an error with the text,
+// without having to read till EOF.
+func TestInvalidSyntaxQuickAbort(t *testing.T) {
+	lbls := []string{"foo", "bar"}
+	firstHalf := genSeries(lbls, 0, 10, 1)
+	invalidMetrics := []tsdb.MetricSample{
+		{
+			Timestamp: -1,
+			Value:     0,
+			Labels:    labels2.FromStrings(" INVALID", "INVALID"),
+		},
+	}
+	secondHalf := genSeries(lbls, 10, 200, 1)
+	combined := append(firstHalf, append(invalidMetrics, secondHalf...)...)
+	importText := genOpenMetricsText("invalid_test", "gauge", combined)
+
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	testutil.Ok(t, err)
+
+	r := strings.NewReader(importText)
+	err = ImportFromFile(r, tmpDbDir, maxSamplesInMemory, maxBlockChildren, nil)
+
+	testutil.NotOk(t, err)
+	// We should abort and return the error with the text, without reading till EOF.
+	testutil.Assert(t, r.Len() > 0, "import text has been completely read")
+}
+
+// func benchVirtual(vThresh int, b *testing.B) {
```

Comment:
Avoid commented, dead code in the PR. It makes this PR looks like it's unfinished :hugs:

Reply:
Yeah I think I meant to remove this before merge.